python/_tskitmodule.c

```diff
@@ -9860,6 +9860,150 @@ TreeSequence_divergence_matrix(TreeSequence *self, PyObject *args, PyObject *kwd
+    ret = 0;
+out:
+    *ret_num_time_windows = num_time_windows;
+    *ret_array = time_windows_array;
+    return ret;
+}
+
+static int
+parse_set_indexes(PyObject *indexes, PyArrayObject **ret_array,
+    tsk_size_t *ret_num_indexes, npy_intp tuple_size)
+{
+    int ret = -1;
+    tsk_size_t num_indexes = 0;
+    PyArrayObject *indexes_array = NULL;
+    npy_intp *shape;
+
+    indexes_array = (PyArrayObject *) PyArray_FROMANY(
+        indexes, NPY_INT32, 2, 2, NPY_ARRAY_IN_ARRAY);
+    if (indexes_array == NULL) {
+        goto out;
+    }
+    shape = PyArray_DIMS(indexes_array);
+    if (shape[0] < 1 || shape[1] != tuple_size) {
+        PyErr_Format(
+            PyExc_ValueError, "indexes must be a k x %d array.", (int) tuple_size);
+        goto out;
+    }
+    num_indexes = shape[0];
+    ret = 0;
+out:
+    *ret_num_indexes = num_indexes;
+    *ret_array = indexes_array;
+    return ret;
+}
+
+static PyObject *
+TreeSequence_pair_coalescence_counts(TreeSequence *self, PyObject *args, PyObject *kwds)
+{
+    PyObject *ret = NULL;
+
+    static char *kwlist[] = { "windows", "sample_set_sizes", "sample_sets", "indexes",
+        "time_windows", "span_normalise", NULL };
+    PyObject *py_sample_set_sizes = Py_None;
+    PyObject *py_sample_sets = Py_None;
+    PyObject *py_windows = Py_None;
+    PyObject *py_time_windows = Py_None;
+    PyObject *py_indexes = Py_None;
+    PyArrayObject *result_array = NULL;
+    PyArrayObject *windows_array = NULL;
+    PyArrayObject *time_windows_array = NULL;
+    PyArrayObject *indexes_array = NULL;
+    PyArrayObject *sample_set_sizes_array = NULL;
+    PyArrayObject *sample_sets_array = NULL;
+    tsk_flags_t options = 0;
+    tsk_size_t num_indexes = 0;
+    tsk_size_t num_sample_sets = 0;
+    tsk_size_t num_windows = 0;
+    tsk_size_t num_time_windows = 0;
+    int span_normalise = 0;
+    int err;
+
+    if (TreeSequence_check_state(self) != 0) {
+        goto out;
+    }
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "OOOOO|i", kwlist, &py_windows,
+            &py_sample_set_sizes, &py_sample_sets, &py_indexes, &py_time_windows,
+            &span_normalise)) {
+        goto out;
+    }
+    if (parse_sample_sets(py_sample_set_sizes, &sample_set_sizes_array, py_sample_sets,
+            &sample_sets_array, &num_sample_sets)
+        != 0) {
+        goto out;
+    }
+    if (parse_windows(py_windows, &windows_array, &num_windows) != 0) {
+        goto out;
+    }
+    if (parse_set_indexes(py_indexes, &indexes_array, &num_indexes, 2) != 0) {
+        goto out;
+    }
+    if (parse_time_windows(py_time_windows, &time_windows_array, &num_time_windows)
+        != 0) {
+        goto out;
+    }
+    if (span_normalise) {
+        options |= TSK_STAT_SPAN_NORMALISE;
+    }
+
+    tsk_size_t num_nodes = tsk_treeseq_get_num_nodes(self->tree_sequence);
+    npy_intp dims[3];
+    dims[0] = num_windows;
+    dims[1] = num_time_windows > 0 ? num_time_windows : num_nodes;
```

Comment:
Yeah, let's insist on passing in explicit time windows. There's nothing to be gained (you malloc the memory later anyway) so let's simplify where we can.